internal/searchindex/searchindex.go

```diff
@@ -0,0 +1,440 @@
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	storedSource, err := convertStoredSource(index.StoredSource)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	return &SearchIndex{
+		SearchIndex: akov2.SearchIndex{
+			Name:           index.Name,
+			DBName:         index.Database,
+			CollectionName: index.CollectionName,
+			Type:           pointer.GetOrDefault(index.Type, ""),
+			Search:         search,
+			VectorSearch:   vectorSearch,
+		},
+		AtlasSearchIndexConfigSpec: akov2.AtlasSearchIndexConfigSpec{
+			Analyzer:       index.Analyzer,
+			Analyzers:      analyzers,
+			SearchAnalyzer: index.SearchAnalyzer,
+			StoredSource:   storedSource,
+		},
+		ID:     index.IndexID,
+		Status: index.Status,
+	}, errors.Join(errs...)
+}
+
+func (s *SearchIndex) cleanup(cleaners ...indexCleaner) {
+	s.ID = nil
+	s.Status = nil
+
+	for _, cleaner := range cleaners {
+		cleaner(s)
+	}
+}
+
+type indexCleaner func(s *SearchIndex)
+
+func cleanConfigRef() indexCleaner {
+	return func(s *SearchIndex) {
+		if s.Search == nil {
+			return
+		}
+		s.Search.SearchConfigurationRef.Name = ""
+		s.Search.SearchConfigurationRef.Namespace = ""
+	}
+}
+
+// This is because API never returns nil
+func cleanVectorSearch() indexCleaner {
+	return func(s *SearchIndex) {
+		if s.VectorSearch == nil {
+			s.VectorSearch = &akov2.VectorSearch{}
+		}
+	}
+}
+
+// This is because API never returns nil
+func cleanSynonyms() indexCleaner {
+	return func(s *SearchIndex) {
+		if s.Search == nil {
+			return
+		}
+		if s.Search.Synonyms == nil {
+			s.Search.Synonyms = &([]akov2.Synonym{})
+		}
+	}
+}
+
+func cleanStoredSource() indexCleaner {
+	return func(s *SearchIndex) {
+		if s.StoredSource == nil {
+			return
+		}
+		if string(s.StoredSource.Raw) == "null" {
+			s.StoredSource = nil
+		}
+	}
+}
+
+func (s *SearchIndex) EqualTo(value *SearchIndex) bool {
+	if value == nil {
+		return false
+	}
+	copySelf := &SearchIndex{
+		SearchIndex:                *s.SearchIndex.DeepCopy(),
+		AtlasSearchIndexConfigSpec: *s.AtlasSearchIndexConfigSpec.DeepCopy(),
+		ID:                         nil,
+		Status:                     nil,
+	}
+	copyValue := &SearchIndex{
+		SearchIndex:                *value.SearchIndex.DeepCopy(),
+		AtlasSearchIndexConfigSpec: *value.AtlasSearchIndexConfigSpec.DeepCopy(),
+		ID:                         nil,
+		Status:                     nil,
+	}
+	copySelf.cleanup(
+		cleanConfigRef(),
+		cleanVectorSearch(),
+		cleanSynonyms(),
+		cleanStoredSource(),
+	)
+	copyValue.cleanup(
+		cleanConfigRef(),
+		cleanVectorSearch(),
+		cleanSynonyms(),
+		cleanStoredSource(),
+	)
+
+	//b, _ := json.MarshalIndent(*copySelf, "", " ")
+	//fmt.Println("AKO INDEX", string(b))
+	//
+	//d := cmp2.Diff(copySelf, copyValue, cmpopts.EquateEmpty())
+	//fmt.Println("INDEX DIFF", d)
+
+	return cmp.SemanticEqual(copySelf, copyValue)
+}
+
+func (s *SearchIndex) Normalize() *SearchIndex {
+	// TODO: Refactor interface to return error!
+	_ = cmp.Normalize(s)
```

Comment:
hmmm ... we really should handle the error here. many things can go wrong unnoticed.

Reply:
I updated the interface and now the error has to be handled by the caller. For the search index it bubbles up to the "update" action and handled accordingly.